### PR TITLE
chore: [Plugin Action Form] Common Editor State

### PR DESCRIPTION
**Pull request by hetunandu**
## Description

Passes the correct states for the Common Editor form in the Plugin Action Form.
- Uses a hook to pass the form states as already implemented
- Uses helper functions to get the header and params count from the said state
    - As a side effect I can remove these from the original implementation since the invocation is now in the common `RequestTabs` components
- Updates the `changeActionCall` hook to only be called if the action changes
   - Updates the tests to reflect this

EE PR to track tests: https://github.com/appsmithorg/appsmith-ee/pull/5217

Parts of #36154 

## Automation

/ok-to-test tags="@tag.Datasource"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11026451216>
> Commit: 2b0537816ae7501062bd6100fde560b034c9708a
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11026451216&attempt=2" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Datasource`
> Spec:
> <hr>Wed, 25 Sep 2024 08:01:34 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Release Notes

- **New Features**
  - Introduced a custom hook for streamlined retrieval of action-related values in forms.
  - Added utility functions for counting valid headers and parameters.

- **Improvements**
  - Simplified component interfaces by removing unnecessary properties related to headers and parameters.
  - Enhanced type safety for header and parameter properties.
  - Refined tab rendering logic for better user experience.

- **Bug Fixes**
  - Adjusted logic to ensure actions are dispatched only on changes to action IDs.

- **Documentation**
  - Updated relevant documentation to reflect changes in component props and functionalities.

### CODE REVIEW

**Review by ashit-rath — APPROVED** [review]

**Comment by ashit-rath** [comment]
@hetunandu Can we address the comments from coderabbitai?

File: app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/hooks/useGetFormActionValues.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  ) as Property[];
+
+  const datasources = useSelector(getDatasources);
+
+  let datasourceFromAction: Action["datasource"] | undefined = get(
+    formValues,
+    "datasource",
+  );
+
+  if (datasourceFromAction && datasourceFromAction.hasOwnProperty("id")) {
```
[the remark this answers is not in the document]
Comment (albinAppsmith) [reply app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/hooks/useGetFormActionValues.ts:36]: @hetunandu this is also recommended by MDN as well
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty

**Review by albinAppsmith — APPROVED** [review]

File: app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/utils/getValidProperties.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import type { Property } from "entities/Action";
+
+function getValidProperties(value?: Property[]) {
+  if (!Array.isArray(value)) {
+    return [];
+  }
+
+  return value.filter((v) => v.key && v.key !== "");
+}
```
[the remark this answers is not in the document]
Comment (hetunandu) [reply app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/utils/getValidProperties.ts:9]: Not needed

File: app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/hooks/useGetFormActionValues.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (hetunandu) [reply app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/hooks/useGetFormActionValues.ts]: Done

**Review by albinAppsmith — APPROVED** [review]

File: app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/utils/getHeadersCount.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  const validDatasourceHeaders = getValidProperties(datasourceHeaders);
+  const validAutoGeneratedHeaders = getValidProperties(
+    autoGeneratedActionHeaders,
+  );
+
+  return (
+    validActionHeaders.length +
+    validDatasourceHeaders.length +
+    validAutoGeneratedHeaders.length
+  );
```
[the remark this answers is not in the document]
Comment (hetunandu) [reply app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/utils/getHeadersCount.ts:19]: Not needed

**Review by ashit-rath — APPROVED** [review]

File: app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/hooks/useGetFormActionValues.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  ) as Property[];
+
+  const autoGeneratedHeaders: ApiAction["actionConfiguration"]["autoGeneratedHeaders"] =
+    get(formValues, "actionConfiguration.autoGeneratedHeaders", []);
+
+  const actionParams = get(
+    formValues,
+    "actionConfiguration.queryParameters",
+    [],
+  ) as Property[];
```
[the remark this answers is not in the document]
Comment (hetunandu) [reply app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/hooks/useGetFormActionValues.ts]: Done

File: app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/hooks/useGetFormActionValues.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  let datasourceFromAction: Action["datasource"] | undefined = get(
+    formValues,
+    "datasource",
+  );
+
+  if (datasourceFromAction && datasourceFromAction.hasOwnProperty("id")) {
+    datasourceFromAction = datasources.find(
+      (d) => d.id === datasourceFromAction?.id,
+    );
+  }
```
[the remark this answers is not in the document]
Comment (hetunandu) [reply app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/hooks/useGetFormActionValues.ts]: Done

File: app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/RequestTabs.tsx [hunk 6]
```diff
@@ -57,27 +49,41 @@ export function RequestTabs(props: {
...
-                    : undefined
-              }
-              value={tab}
-            >
-              {createMessage(API_EDITOR_TAB_TITLES[tab])}
-            </Tab>
-          ))}
+          {Object.values(API_EDITOR_TABS)
+            .filter((tab) => {
+              return !(!props.showSettings && tab === API_EDITOR_TABS.SETTINGS);
```
Comment (ayushpahwa) [inline app/client/src/PluginActionEditor/components/PluginActionForm/components/CommonEditorForm/RequestTabs.tsx:69]: nit: consider adding a comment here for better understanding. (Not a blocker, can be picked up in separate PR)

**Review by ayushpahwa — APPROVED** [review]

**Review by ankitakinger — APPROVED** [review]
